File: tutorials/publish_subscribe2/c++11/home_automation_publisher.cxx
//
// (c) 2024 Copyright, Real-Time Innovations, Inc.  All rights reserved.
//
//  RTI grants Licensee a license to use, modify, compile, and create derivative
//  works of the Software.  Licensee has the right to distribute object form
//  only for use with RTI products.  The Software is provided "as is", with no
//  warranty of any type, including any warranty for fitness for any purpose.
//  RTI is under no obligation to maintain or support the Software.  RTI shall
//  not be liable for any incidental or consequential damages arising out of the
//  use or inability to use the software.
//

#include <iostream>
#include <thread>
#include <rti/rti.hpp>
#include "home_automation.hpp"

void publish_sensor(
    const std::string& sensor_name,
    const std::string& room_name)
{
    dds::domain::DomainParticipant participant(0);
    dds::topic::Topic<DeviceStatus> topic(participant, "WindowStatus");
    dds::pub::DataWriter<DeviceStatus> writer(topic);

    DeviceStatus device_status { sensor_name, room_name, false };
    for (int i = 0; i < 1000; i++) {
        device_status.is_open(!device_status.is_open());
        writer.write(device_status);
        std::this_thread::sleep_for(std::chrono::seconds(2));
    }
}

int main(int argc, char **argv)
{
    std::string sensor_name = (argc > 1) ? argv[1] : "Sensor1";
    std::string room_name = (argc > 2) ? argv[2] : "LivingRoom";
    publish_sensor(sensor_name, room_name);
}

File: tutorials/distributed_data_cache/cs/HomeAutomationPublisher.cs
//
// (c) 2024 Copyright, Real-Time Innovations, Inc.  All rights reserved.
//
//  RTI grants Licensee a license to use, modify, compile, and create derivative
//  works of the Software.  Licensee has the right to distribute object form
//  only for use with RTI products.  The Software is provided "as is", with no
//  warranty of any type, including any warranty for fitness for any purpose.
//  RTI is under no obligation to maintain or support the Software.  RTI shall
//  not be liable for any incidental or consequential damages arising out of the
//  use or inability to use the software.
//

using System.Threading;
using Rti.Dds.Domain;
using Rti.Dds.Publication;
using Rti.Dds.Topics;

namespace HomeAutomation
{
    public class Publisher
    {
        public static void PublishSensor(string sensorName, string roomName)
        {
            DomainParticipant participant =
                DomainParticipantFactory.Instance.CreateParticipant(domainId: 0);
            Topic<DeviceStatus> topic = participant.CreateTopic<DeviceStatus>("WindowStatus");
            DataWriter<DeviceStatus> writer =
                participant.ImplicitPublisher.CreateDataWriter(topic);

            // Create a DeviceStatus sample
            var deviceStatus = new DeviceStatus
            {
                sensor_name = sensorName,
                room_name = roomName,
                is_open = false
            };

            for (int i = 0; i < 20; i++)
            {
                // Simulate the window opening and closing
                deviceStatus.is_open = !deviceStatus.is_open;
                writer.Write(deviceStatus);
                Thread.Sleep(2000); // Sleep for 2 seconds
            }
        }
    }
}
